index.rst

```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+
+Per-HDU Cells
+-------------
+
+Another simple file layout is to put each image plane for each cell in a completely separate FITS image HDU.
+This is entirely compatible with FITS tile compression (though we'd almost certainly compress the entire HDU as one tile) and our goals for using FITS WCS.
+Stitching images from different HDUs into a coherent whole is probably a bit more likely for a third-party FITS viewer to support than images from different binary tables, but a flat list of HDUs for all cells and image planes provides a lot less organizational structure than a binary table (especially a single binary table) for third-party tools to interpret.
+
+Each HDU comes with an extra 3-9 KB of overhead (1-2 header blocks, and padding out the full HDU size to a multiple of 2880 bytes) that cannot be compressed, which is not ideal, but probably not intolerable unless we get unexpectedly good compression ratios or shrink the cell size: an uncompressed 250×250 single-precision floating point image is 250KB, so those overheads should be at most 4% or so.
+The overheads would be significant for the PSF images, which we expect to be 25-40 pixels on a side (2.5-6 KB uncompressed).
+
+Subimage reads would be similarly non-ideal but perhaps tolerable.
+Because each HDU is so small, it'd be plenty efficient to read full HDUs, but only those for the cells that overlap the region of interest.
+Seeking to the right HDUs (or requesting the appropriate byte ranges, in the object store case) is easily solved by putting a table of byte offsets in the primary HDU header, though this isn't something third-party FITS readers could leverage.
+That would make for a simple solution to the problem of doing subimage reads over object stores (including compression): we could use the address table to read the HDUs we are interested in in their entirety into a client-side memory location that looks like a full in-memory FITS file holding just those HDUs, and then delegate to CFITSIO's "memory file" interfaces to let it do the decompression.
+
+As in the binary table case, it's an open question whether we could get sufficiently good compression ratios if we are limited to compressing one cell at a time.
+
+Data Cubes
+----------
+
+In this layout, we'd have one 3-d or 4-d image extension HDU for each plane, with each cell's image a 2-d slice of that higher-dimensional array, and the other dimensions corresponding to a 1-d or 2-d index of that cell in its grid.
+
+This avoids the problem with per-HDU overheads, and it makes an address table much less important, as there are many fewer HDUs.
+It also allows compression tiles that comprise multiple cells.
+
+It does not allow us to represent the on-sky locations of cells using FITS WCS, however, and this is probably enough to rule it out.
```

Comment:
Are you referring to the integer offsets that indicate the position of the cell?

Reply:
Yes, more or less.  We could have a WCS that describes one cell, but I don't think there's any way to write a FITS WCS that puts different slices of a data cube in different places.

Reply:
I don't think FITS-WCS can do it. I'm pretty sure AST can with switchMaps.

Reply:
No idea if gwcs can.

Reply:
I believe there might be a way to do this with table-lookup WCSes, but they are not uniformly well supported by readers.  I would have to read the FITS standard very closely to determine whether this would also require us to do all the spherical projection ourselves, which would require us to be careful about precision issues in the (piecewise linear) table lookups, especially near the coordinate poles.

I am not endorsing this idea!  But if this storage scheme were the winner for other reasons, it's something we could research.

Reply:
This storage scheme definitely isn't winning for other reasons, so I don't think it's worth anyone's time to look into it further.